File: Object-C/BlackFireIM/Example/ChatDetail/ChatRoom/BFChatRoomEditController.m
//
//  BFChatRoomEditController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/11/7.
//

#import "BFChatRoomEditController.h"
#import "MSIMSDK-UIKit.h"
#import "BFChatRoomMemberListController.h"
#import "BFEditTodInfoController.h"


@interface BFChatRoomEditController ()<UITableViewDelegate,UITableViewDataSource>

@property(nonatomic,strong) UITableView *myTableView;

@end

@implementation BFChatRoomEditController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"Chat Room Setting";
    [self.view addSubview:self.myTableView];
    
    UIButton *quitBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [quitBtn setTitle:@"退出群聊天" forState:UIControlStateNormal];
    [quitBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    quitBtn.titleLabel.font = [UIFont systemFontOfSize:16];
    quitBtn.backgroundColor = [UIColor redColor];
    quitBtn.frame = CGRectMake(0, 0, Screen_Width, 60);
    [quitBtn addTarget:self action:@selector(quitBtnClick) forControlEvents:UIControlEventTouchUpInside];
    self.myTableView.tableFooterView = quitBtn;
}

- (UITableView *)myTableView
{
    if (!_myTableView) {
        _myTableView = [[UITableView alloc]initWithFrame:CGRectMake(0, 0, Screen_Width, Screen_Height) style:UITableViewStylePlain];
        _myTableView.dataSource = self;
        _myTableView.delegate = self;
        _myTableView.rowHeight = 60;
        _myTableView.contentInset = UIEdgeInsetsMake(NavBar_Height + StatusBar_Height, 0, Bottom_SafeHeight, 0);
        _myTableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        [_myTableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"cell"];
    }
    return _myTableView;
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 4;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    if (indexPath.row == 0) {
        cell.textLabel.text = @"Member List";
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%zd 人",self.roomInfo.onlineCount];
    }else if (indexPath.row == 1) {
        cell.textLabel.text = @"Chat room name";
        cell.detailTextLabel.text = self.roomInfo.room_name;
    }else if (indexPath.row == 2) {
        cell.textLabel.text = @"Tips of day";
    }else if (indexPath.row == 3) {
        cell.textLabel.text = self.roomInfo.is_mute ? @"Cancel Mute" : @"Mute";
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (indexPath.row == 0) {
        BFChatRoomMemberListController *vc = [[BFChatRoomMemberListController alloc]init];
        vc.roomInfo = self.roomInfo;
        [self.navigationController pushViewController:vc animated:YES];
    }else if (indexPath.row == 1) {//修改聊天室名称
        [MSHelper showToastString:@"聊天室名称只支持后台修改"];
    }else if (indexPath.row == 2) {//修改公告
        if (self.roomInfo.action_tod == NO) {
            [MSHelper showToastString:@"只有管理员才能发布公告"];
            return;
        }
        BFEditTodInfoController *vc = [[BFEditTodInfoController alloc]init];
        vc.roomInfo = self.roomInfo;
        [self.navigationController pushViewController:vc animated:YES];
    }else if (indexPath.row == 3) {
        
        [self editChatRoomMuteStatus];
        
    }
}

- (void)editChatRoomMuteStatus
{
    WS(weakSelf)
    [[MSIMManager sharedInstance] muteChatRoom:!self.roomInfo.is_mute toRoom_id:self.roomInfo.room_id duration:1 successed:^{
        
        [MSHelper showToastSucc:@"Success"];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [weakSelf.myTableView reloadData];
        });
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
}

- (void)quitBtnClick
{
    [self.navigationController popToRootViewControllerAnimated:YES];
}

@end

File: Swift/MSIM-Swift/ChatDetail/ChatRoom/BFChatRoomEditController.swift
//
//  BFChatRoomEditController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/11/9.
//

import UIKit
import MSIMSDK


public class BFChatRoomEditController: BFBaseViewController {

    public var roomInfo: MSGroupInfo!
    
    private lazy var myTableView: UITableView = {
        let tableView = UITableView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.height), style: .plain)
        tableView.dataSource = self
        tableView.delegate = self
        tableView.rowHeight = 60
        tableView.contentInset = UIEdgeInsets(top: UIScreen.status_navi_height, left: 0, bottom: UIScreen.safeAreaBottomHeight, right: 0)
        tableView.contentInsetAdjustmentBehavior = .never
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        return tableView
    }()
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "Chat Room Setting"
        view.addSubview(myTableView)
        
        let quitBtn = UIButton(type: .custom)
        quitBtn.setTitle("退出群聊", for: .normal)
        quitBtn.setTitleColor(.white, for: .normal)
        quitBtn.titleLabel?.font = .systemFont(ofSize: 16)
        quitBtn.backgroundColor = .red
        quitBtn.frame = CGRect(x: 0, y: 0, width: UIScreen.width, height: 60)
        quitBtn.addTarget(self, action: #selector(quitBtnClick), for: .touchUpInside)
        self.myTableView.tableFooterView = quitBtn
    }
    
    @objc func quitBtnClick() {
        navigationController?.popToRootViewController(animated: true)
    }
}

extension BFChatRoomEditController: UITableViewDataSource,UITableViewDelegate {
    
    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if self.roomInfo.action_tod {
            return 4
        }
        return 3
    }
    
    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        cell.accessoryType = .disclosureIndicator
        if indexPath.row == 0 {
            cell.textLabel?.text = "Member List"
            cell.detailTextLabel?.text = "\(self.roomInfo.onlineCount) 人"
        }else if indexPath.row == 1 {
            cell.textLabel?.text = "Chat room name"
            cell.detailTextLabel?.text = self.roomInfo.room_name
        }else if indexPath.row == 2 {
            cell.textLabel?.text = "Tips of day"
        }else if indexPath.row == 3 {
            cell.textLabel?.text = self.roomInfo.is_mute ? "Cancel Mute" : "Mute"
        }
        return cell
    }
    
    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.row == 0 {
            let vc = BFChatRoomMemberListController()
            vc.roomInfo = self.roomInfo
            navigationController?.pushViewController(vc, animated: true)
        }else if indexPath.row == 1 {//修改聊天室名称
            MSHelper.showToastWithText(text: "聊天室名称只支持后台修改")
        }else if indexPath.row == 2 {//修改公告

            let vc = BFEditTodInfoController()
            vc.roomInfo = self.roomInfo
            self.navigationController?.pushViewController(vc, animated: true)
            vc.editComplete = {[weak self] in
                self?.myTableView.reloadData()
            }
        }else if indexPath.row == 3 {
            editChatRoomMuteStatus()
        }
    }
    
    private func editChatRoomMuteStatus() {
        MSIMManager.sharedInstance().muteChatRoom(!self.roomInfo.is_mute, toRoom_id: self.roomInfo.room_id, duration: 1) {[weak self] in
            
            MSHelper.showToastSuccWithText(text: "Success")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self?.myTableView.reloadData()
            }
        } failed: { _, desc in
            MSHelper.showToastFailWithText(text: desc ?? "")
        }
    }
}
